### PR TITLE
Remove all references to Person from EMPI.

### LINKED ISSUE #2161: Support Arbitrary resource MDM
Current EMPI implementation is useful for Patients/Practitioners, but there is large demand for Master Data Management over arbitrary resource types, e.g. medication, organization. 

To this end we should: 
1. Drop references to person, and instead rely directly on MPI_LINK table results. 
2. Replace usages of person with an instance of the same MDM'ed class. e.g. Patient links to Patient, Organization links to Organization. Note these links are now only maintained in the MPI_LINK table. 
3. Deprecate some of the MPI_LINK columns
4. Add MDM toggles on a per-resource basis. Maybe in the empi-rules.json itself?
5. Auto-create subscriptions based on these new MDM types.

### PR DESCRIPTION
**Pull request by tadgh**
Closes #2161. 

Many large scale changes. The primary purpose of this PR is to change the EMPI solution to an MDM solution. In broad strokes, this means that the `Person` resource is no longer special. 

The new solution (MDM) involves creating another resource where Person used to be used. For example, if MDM is occurring on a patient, we will create a new Patient for them, and tag that patient as a `Golden Record`. This means that several things ahve changed:
1. Provider methods that used to point to type of `Person` are now server-level operations in which you specify a resource type. 
2. Link updating and querying no longer rely on Person IDs, but instead on arbitrary resource ids, depending on the resource type youre referring to in MDM. 
3. Change to the EMPI config to require a list of `mdmTypes`

### CODE REVIEW

**Comment by nvg-smile** [comment]
Looks great to me, only a few minor changes to consider for the future. We may need to review comments, as they still have many mentions of patient/person. Naming is sometimes inconsistent. Any mention of person and source resource should be updated to golden resource. Considering this is still WIP, we may address those in the future.

**Review by nvg-smile — CHANGES_REQUESTED** [review]
Great work :)! For the most part - naming convention changes. Exciting to see so much being accomplished so quickly!

**Review by nvg-smile — COMMENTED** [review]
EMPI to MDM references should be addressed

**Comment by bdenton** [comment]
Can I please suggest that the `ca.uhn.fhir.rest.server` Java package please be renamed? This package already exists in `hapi-fhir-server` and it creates a  "split package" situation that is a problem in many environments.. notable any sort of OSGi class loading implementation.
cheers, Bill Denton

**Review by nvg-smile — APPROVED** [review]

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/interceptor/MdmStorageInterceptor.java [hunk 1]
```diff
@@ -45,72 +45,71 @@
...
+		if (myMdmSettings.isPreventMultipleEids()) {
+			forbidIfHasMultipleEids(theUpdatedResource);
 		}
 
-		if (EmpiUtil.isEmpiManagedPerson(myFhirContext, theNewResource) &&
-			myPersonHelper.isDeactivated(theNewResource)) {
-			ourLog.debug("Deleting empi links to deactivated Person {}", theNewResource.getIdElement().toUnqualifiedVersionless());
-			int deleted = myEmpiLinkDeleteSvc.deleteNonRedirectWithWithAnyReferenceTo(theNewResource);
+		//TODO GGG MDM: Check if this is actually handled already in mdm update code or not.
+		if (myGoldenResourceHelper.isDeactivated(theUpdatedResource)) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/interceptor/MdmStorageInterceptor.java:88]: I was initially thrown by the term "deactivated" but I think it's a good term here.  It does make me wonder if there's a better one.  What do other mdm systems call this...?

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmResourceFilteringSvc.java [hunk 2]
```diff
@@ -32,43 +33,48 @@
...
+	 * sufficient information to perform meaningful MDM processing. (For example, how can MDM processing occur on a patient that has _no_ attributes?)
 	 *
-	 * @param theResource the resource that you wish to check against EMPI rules.
+	 * @param theResource the resource that you wish to check against MDM rules.
 	 *
-	 * @return whether or not EMPI processing should proceed
+	 * @return whether or not MDM processing should proceed
 	 */
 	public boolean shouldBeProcessed(IAnyResource theResource) {
+		//TODO GGG ask KHS: Skip the infinite loop, whoops. Better way to do this? tighter subscription criteria?
```
Comment (fil512) [inline hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/MdmResourceFilteringSvc.java:59]: hmm.  I don't understand the question

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/candidate/FindCandidateByScoreSvc.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class FindCandidateByScoreSvc extends BaseCandidateFinder {
```
Comment (fil512) [inline hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/candidate/FindCandidateByScoreSvc.java:44]: Consider renaming to `FindCandidateByExampleSvc`

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/model/CanonicalIdentityAssuranceLevel.java [hunk 4]
```diff
@@ -1,8 +1,8 @@
-package ca.uhn.fhir.empi.model;
+package ca.uhn.fhir.mdm.model;
 
 /*-
  * #%L
- * HAPI FHIR - Enterprise Master Patient Index
+ * HAPI FHIR - Master Data Management
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/model/CanonicalIdentityAssuranceLevel.java:5]: delete this class

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/AssuranceLevelUtil.java [hunk 5]
```diff
@@ -20,21 +20,21 @@
...
+import ca.uhn.fhir.mdm.api.MdmMatchResultEnum;
+import ca.uhn.fhir.mdm.model.CanonicalIdentityAssuranceLevel;
 import ca.uhn.fhir.rest.server.exceptions.InvalidRequestException;
 
 /**
  * Helper class to determine assurance level based on Link Source and Match Result.
- * This is strictly for use in populating Person links.
+ * This is strictly for use in populating Golden Resource links.
  */
 public final class AssuranceLevelUtil {
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/AssuranceLevelUtil.java:32]: Delete all mdm classes concerning assurance level

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/GoldenResourceHelper.java [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+			BaseRuntimeElementCompositeDefinition<?> compositeDefinition = (BaseRuntimeElementCompositeDefinition<?>) childDefinition.getChildByName(field);
+			IBase newFieldValue = compositeDefinition.newInstance();
+			terser.cloneInto(theFromFieldValue, newFieldValue, true);
+
+			theToFieldValues.add(newFieldValue);
+		}
+	}
+
+	private boolean contains(IBase theItem, List<IBase> theItems) {
+		PrimitiveTypeComparingPredicate predicate = new PrimitiveTypeComparingPredicate();
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/GoldenResourceHelper.java:171]: Is this actually a `PrimitiveTypeEqualsPredicate` ?

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/GoldenResourceHelper.java [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+					cloneEidIntoResource(theGoldenResourceIdentifier, base, theNewGoldenResource);
+					ourLog.debug("System {} differs from system in the MDM rules {}", baseSystem, mdmSystem);
+				}
+			} else {
+				ourLog.debug("System is missing, skipping");
+			}
+		}
+	}
+
+	private void validateContextSupported() {
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/GoldenResourceHelper.java:198]: This should probably be a static util function so we can use it elsewhere..

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/GoldenResourceHelper.java [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	public void deactivateResource(IAnyResource theResource) {
+		MdmUtil.setGoldenResourceRedirected(theResource);
+	}
+
+	public boolean isDeactivated(IBaseResource theGoldenResource) {
+		return MdmUtil.isGoldenRecordRedirected(theGoldenResource);
+	}
+}
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/GoldenResourceHelper.java:377]: nitpick this class is getting kind of long.  is there a distinct responsibility in here that could be carved out into a separate class?

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/MdmUtil.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+ */
+
+import ca.uhn.fhir.mdm.api.MdmConstants;
+import org.hl7.fhir.instance.model.api.IBaseCoding;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+
+import javax.annotation.Nonnull;
+import java.util.Optional;
+
+public final class MdmUtil {
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/MdmUtil.java:30]: MdmResourceUtil?

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/PrimitiveTypeComparingPredicate.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+ * limitations under the License.
+ * #L%
+ */
+
+import org.hl7.fhir.instance.model.api.IPrimitiveType;
+
+import java.lang.reflect.Field;
+import java.util.function.BiPredicate;
+
+public class PrimitiveTypeComparingPredicate implements BiPredicate {
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/PrimitiveTypeComparingPredicate.java:28]: nitpick: rename to `PrimitiveTypeEqualsPredicate`

File: hapi-fhir-server-mdm/src/test/java/ca/uhn/fhir/mdm/util/PrimitiveTypeComparingPredicateTest.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+		myTerser = myFhirContext.newTerser();
+
+		myPositiveTest1 = newPatient();
+		myPositiveTest2 = newPatient();
+		myPositiveTest3 = newPatient();
+
+		Patient patient = newPatient();
+		patient.setActive(false);
+		patient.setMultipleBirth(new BooleanType(false));
+		myNegativeTest = patient;
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/test/java/ca/uhn/fhir/mdm/util/PrimitiveTypeComparingPredicateTest.java:53]: why is there a Patient patient?

File: hapi-fhir-server-mdm/src/test/java/ca/uhn/fhir/mdm/util/PrimitiveTypeComparingPredicateTest.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+	@Test
+	public void testReflexivity() {
+		assertTrue(cut.test(myPositiveTest1, myPositiveTest1));
+	}
+
+	@Test
+	public void testTransitivity() {
+		assertTrue(cut.test(myPositiveTest1, myPositiveTest2));
+		assertTrue(cut.test(myPositiveTest2, myPositiveTest3));
+		assertTrue(cut.test(myPositiveTest1, myPositiveTest3));
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/test/java/ca/uhn/fhir/mdm/util/PrimitiveTypeComparingPredicateTest.java:100]: Nitpick: This test is misleading.  It's not testing transitivity.  It's just testing that three identical resources are all equal to each other.  It doesn't do anything beyond testing that `assertTrue(cut.test(myPositiveTest1, myPositiveTest2));`

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/GoldenResourceHelper.java [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+
+	/**
+	 * Clones specified composite field (collection). Composite field values must confirm to the collections
+	 * contract.
+	 *
+	 * @param theFrom Resource to clone the specified filed from
+	 * @param theTo   Resource to clone the specified filed to
+	 * @param field   Field name to be copied
+	 */
+	private void cloneCompositeField(IBaseResource theFrom, IBaseResource theTo, String field) {
```
Comment (fil512) [inline hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/util/GoldenResourceHelper.java:147]: This is really nice!  Will be interesting to see how it evolves as we add survivability rules!

**Review by fil512 — APPROVED** [review]

**Pull request merged**